### ISSUE TITLE
Why am I throttled even the hour total request count is far below than 30k?

### ISSUE THREAD

**Issue by lissdy** [Azure/AppConfiguration#736]
![image](https://user-images.githubusercontent.com/3861170/228499563-701eb849-3009-4f33-8c45-6f5603c21663.png)
As the chart shows, why am I throttled even though the hour total request count(2k) is far below than 30k?

**Comment by zhenlan** [comment]
It's likely due to the burst of requests in a very short time. Please check out the doc below.

https://learn.microsoft.com/en-us/azure/azure-app-configuration/faq#my-application-receives-http-status-code-429-responses--why
You'll receive an HTTP status code 429 response under these circumstances:

- Exceeding the daily request limit for a store in the Free tier.
- Exceeding the hourly request limit for a store in the standard tier.
- **Momentary throttling due to a large burst of requests†.**
- Excessive bandwidth usage.
- Attempting to create or modify a key when the storage quota is exceeded.

**Comment by lissdy** [comment]
I am trying to import configuration files via pipeline, may I know any limitations for configuration file import? How many configuration items can be included in one file? And May I know how to check how much storage has been used for one store? Thanks

**Comment by lissdy** [comment]
And if  App Service failed to connect DB, that would cause the app service to keep starting up. Will that lead to App Configuration throttling? Thanks @zhenlan

**Comment by zhenlan** [comment]
@lissdy, as mentioned in the same document, the momentary throttling can often be ignored as App Configuration clients have built-in retry logic. I would encourage you to try it out and see it causes any real issues.

_†A configuration store may experience momentary throttling if it receives a large burst of requests. App Configuration clients, such as the Azure SDK, configuration provider libraries, and Azure Pipeline tasks, automatically retry on throttled requests. For any applications using one of these clients, or a custom client that retries on throttled requests, this momentary throttling should go unnoticed, should it occur._

**Comment by zhenlan** [comment]
> May I know how to check how much storage has been used for one store?

You can check out the "Daily Storage Usage" metrics. Note that it may report zero percent if your usage is not too much more than a few megabytes.
https://learn.microsoft.com/en-us/azure/azure-app-configuration/monitor-app-configuration?tabs=portal#analyzing-metrics

**Comment by lissdy** [comment]
Thanks @zhenlan for the guide, I have emulated the "App Service failed to connect DB" case and Azure App Configuration works perfectly in the scenario!

One more question about monitoring:
How can we monitor our quota limit effectively to prevent throttling? (by avoiding configuration changes in advance) Considering the discussion on https://github.com/Azure/AppConfiguration/issues/716, where it was noted that HTTP requests can highly exceed the hourly quota limit, what other metrics can we use to ensure that we stay within the quota limit and avoid throttling?

**Comment by drago-draganov** [comment]
@lissdy, Azure AppConfiguration Service should provide the appropriate current quota metric. It's a work item that we are taking. Thanks for your request!

**Comment by lissdy** [comment]
Thanks @drago-draganov  and @zhenlan . Another question about the monitor is that since we use App Configuration store among multiple microservices. Is there any way to monitor the HTTP request count/quota limit consumed by each microservice?

**Comment by drago-draganov** [comment]
AppConfiguration service doesn't have knowledge of specific configuration topology or usage relation. Our telemetry is based on HTTP intrinsic parameters. Perhaps you can filter [AACHttpRequest](https://learn.microsoft.com/en-us/azure/azure-monitor/reference/tables/aachttprequest) logs by ClientIPAddress, specific RequestURI form, etc.
If your application set [DiagnosticsOptions.ApplicationId](https://learn.microsoft.com/en-us/dotnet/api/azure.core.diagnosticsoptions.applicationid?view=azure-dotnet#azure-core-diagnosticsoptions-applicationid) for each micro-service, you could also filter by User-Agent.

**Comment by lissdy** [comment]
Thanks, @drago-draganov, @zhenlan  specific RequestURI helps in my case. 

One more question for the App Configuration Http incoming request count metrics, we are following the [primary+secondary App Configuration store pattern](https://learn.microsoft.com/en-us/azure/azure-app-configuration/concept-disaster-recovery?tabs=core2x#failover-between-configuration-stores) for high availability. Since services retrieve the same set of configuration data from two App Configuration stores simultaneously, per my understanding, the Http Incoming request count metrics for the primary AAC store and secondary AAC store should almost the same. But as below chart, I've noticed that the Http Incoming request count metrics for the secondary AAC store are consistently higher than those for the primary AAC store, even though both stores are being accessed simultaneously by services. Can you explain why this might be happening? Appreciate your patience.

![image](https://user-images.githubusercontent.com/3861170/234024661-0525b06e-a60e-4c00-9039-a8b879e3ea85.png)

**Comment by zhenlan** [comment]
Hi @lissdy, geo-replication is the recommended way for enhanced resiliency to regional outages. The document you referenced will be updated shortly. The App Configuration providers have a built-in failover mechanism. Please follow the document below to set it up.

https://learn.microsoft.com/en-us/azure/azure-app-configuration/howto-geo-replication?tabs=dotnet#use-replicas

**Comment by lissdy** [comment]
Hi @zhenlan, we adopted the primary + secondary pattern because geo-replication is not generally available before. May I know if the feature is GA now? If yes we will migrate to geo-replication.
But I am still not sure why the Http Incoming request count metrics for the secondary AAC store are consistently higher than those for the primary AAC store.

**Comment by zhenlan** [comment]
Yes, the geo-replication is GA'ed. See https://github.com/Azure/AppConfiguration-Announcements/issues/29.

Regarding the two-store setup, did you set up refresh for one store but not the other? Do you have other apps using one store but not the other?

**Comment by lissdy** [comment]
Thanks @zhenlan for replying, the refresh is for both. And the two stores shared the same apps. So per my understanding, the two stores' metrics should look similar.

**Comment by zhenlan** [comment]
Yes, I have the same expectation. If you enable logs of your stores, maybe they can give you a clue about what they are.

**Comment by lissdy** [comment]
Thanks @zhenlan , I enabled the logs of configuration stores and noticed something unexpected.

I have set multiple sentinel keys, for any sentinel change, the configuration provider should load the selected configurations. 

![image](https://user-images.githubusercontent.com/3861170/234268396-8e7487d9-6240-4114-ba16-3dc22af363c3.png)

But I check the service log AACHttpRequest and noticed the first sentinel key refresh frequency works as expected, sending HTTP request every 2 mins, but the second sentinel refresh frequency is lower than the first one, and the last one is even lower than the second one.

All sentinel keys exist in the configuration store.

![image](https://user-images.githubusercontent.com/3861170/234498431-a2143bea-8a88-45a2-b456-ab90ee74a462.png)


Sometimes I would only update the last sentinel key, this makes the configuration change take effects much longer than expected.

It looks like a bug to me, may I confirm if is this expected?

version: 5.1.0

leave a comment to the configuration provider repo as well https://github.com/Azure/AppConfiguration-DotnetProvider/issues/407, not sure if it's a question for the SDK or the configuration store

**Comment by zhenlan** [comment]
@lissdy do you mind upgrading to version 6.0.0? You will need that version for the geo-replication too. Version 5.1.0 was released quite a while ago. I just want to make sure we are not chasing a bug that has been addressed already.

**Comment by zhenlan** [comment]
BTW, we just published our updated [Resiliency and Disaster Recovery](https://learn.microsoft.com/en-us/azure/azure-app-configuration/concept-disaster-recovery) guidance document.

**Comment by lissdy** [comment]
We plan to upgrade eventually, but can't finish the process immediately.

**Comment by zhenlan** [comment]
Understood. You can review the [release notes ](https://github.com/Azure/AppConfiguration/blob/main/releaseNotes/MicrosoftExtensionsConfigurationAzureAppConfiguration.md)to see if there are any concerns about the upgrade. As part of the 6.0.0 release, we added richer [logging support](https://learn.microsoft.com/en-us/azure/azure-app-configuration/enable-dynamic-configuration-aspnet-core?tabs=core6x#logging-and-monitoring), which can be helpful for either troubleshooting during the development or monitoring in production.

**Comment by zhenlan** [comment]
@lissdy I hope all your questions are answered. Closing.
